client/src/components/Users/index.jsx

```diff
@@ -1,9 +1,237 @@
+      const nameOptions = filter(allUsers);
+      this.setState({ nameOptions });
+    }
+  }
+  handleAddUser = (event, columnName) => {
+    const newValue = event.target.value;
+    this.setState(prevState => {
+      const clonedUsers = JSON.parse(JSON.stringify(prevState.users));
+      clonedUsers[clonedUsers.length - 1][columnName] = newValue;
+      //Not sure if I should update the state now or after inserting the user in database!
+      return { users: clonedUsers, newRow: clonedUsers[clonedUsers.length - 1], showSaveButton: true };
+    });
+  };
+
+  saveNewUser = () => {
+    if (this.validateUserInfo(this.state.newRow)) {
+      if (this.state.password) {
+        //fetch to add user in the database
+        //change message
+        this.setState({ passwordError: false, saved: true });
+      } else {
+        this.setState({ passwordError: true });
+      }
+    }
+  };
+
+  handleEditUserInfo = (event, record, columnName) => {
+    if (record.id === this.state.users.length) {
+      return this.handleAddUser(event, columnName);
+    }
+    this.setState({ saving: true, saved: false });
+    const newValue = event.target.value;
+    const memberId = record.id;
+
+    // Editing UserInfo
+    const { users } = this.state;
+    const clonedUsers = JSON.parse(JSON.stringify(users));
+    const updatedUser = clonedUsers.find(user => user.id === memberId);
+    updatedUser[columnName] = newValue;
+
+    //Validate
+    if (this.validateUserInfo(updatedUser)) {
+      this.updateUserInfo(updatedUser, clonedUsers);
+    }
+  };
+
+  handleDeleteUser = event => {
+    const { users, rowSelected } = this.state;
+    const deletedRow = users.filter(user => user.id === rowSelected)[0];
+    this.showSwal(deletedRow);
+  };
+
+  showSwal = deleteMember => {
+    deleteSwal().then(response => {
+      if (response.value) this.confirmDelete(deleteMember);
+    });
+  };
+
+  confirmDelete = deleteMember => {
+    //Fetch deleteMember to delete from datatabase
+    //Update state//
+  };
+
+  validateUserInfo = user => {
+    this.setState({ fullNameError: false, userNameError: false });
+    if (user['user_name'].length < 3) {
+      this.setState({ userNameError: true });
+      return false;
+    }
+    if (user['full_name'].length < 6) {
+      this.setState({ fullNameError: true });
+      return false;
+    }
+    this.setState({ userNameError: false, fullNameError: false });
+    return true;
+  };
+
+  updateUserInfo = (user, users) => {
+    //Fetch to update userInfo//
+    //Then update users in the state//
+    //change message//
+    this.setState({ users, saving: false, saved: true });
+  };
+
+  handleRow = id => {
+    this.setState({ rowSelected: id });
+  };
+
+  showForm = () => {
+    this.setState({ show: true });
+  };
+
+  AddPassword = password => {
+    this.setState({ password, show: false });
+  };
+  cancel = () => {
+    this.setState({ show: false })
+  }
+
+  render() {
+    const columns = [
```

Comment:
This way each time the `render()` function called you'll declare the columns again which is fine if the columns are dynamic.
In our case, the columns ("the header of the table") are static except the last one which declares the dropDown menu

so I suggest you move the declaration of the columns in the class body without the last column
and then add the last column only inside the render function, something like this :
```jsx
this.columns[3] = {
        render: props => {
          return (
            <Dropdown
              key={props.id}
              trigger={['click']}
              rowId={props.id}
              overlay={
                <UserMenu
                  rowId={props.id}
                  submitPassword={this.handleAddPassword}
                  usersLength={this.state.users.length}
                  handleDeleteUser={this.handleDeleteUser}
                  showPasswordPopup={this.showForm}
                />
              }
            >
              <Icon onClick={() => this.handleRow(props.id)} title="click" className="user__ellipsis" type="ellipsis" />
            </Dropdown>
          );
        },
      };
```

As I said, the way you did it works just fine, but I think it'd be better to declare only what you need to change when the state changes inside the render.

Reply:
Why do you think the dropdown is the only column that is dynamic? 
Nothing static in this table. Inputs are editable